general/device.hpp

```diff
@@ -289,6 +289,68 @@ class Device
    { Get().mpi_gpu_aware = force; }
 
    static bool GetGPUAwareMPI() { return Get().mpi_gpu_aware; }
+
+   /** Query the device driver for what memory type a given @a ptr is allocated with. */
+   static MemoryType QueryMemoryType(void* ptr);
+
+   /** @brief The number of hardware compute units/streaming multiprocessors available on a given compute device @a dev. */
+   static int MFEM_HOST_DEVICE NumMultiprocessors(int dev)
+   {
+#if defined(MFEM_USE_CUDA)
+      int res;
+      cudaDeviceGetAttribute(&res, cudaDevAttrMultiProcessorCount, dev);
+      return res;
+#elif defined(MFEM_USE_HIP)
+      int res;
+      hipDeviceGetAttribute(&res, hipDeviceAttributeMultiprocessorCount, dev);
+      return res;
+#else
+      // not compiled with GPU support
+      return 0;
+#endif
+   }
+
+   static int MFEM_HOST_DEVICE NumMultiprocessors()
```

Comment:
I'd move this implementation and the implementation of the other new methods to the `.cpp` file.